### PR TITLE
Feature yang changes

### PR DESCRIPTION
**Pull request by arlakshm**
Signed-off-by: Arvindsrinivasan Lakshmi Narasimhan <arlakshm@microsoft.com>

#### Why I did it
Add yang model for Feature configuration

#### How I did it
Add `feature.yang` and unit tests

#### How to verify it
Compile `target/python-wheels/sonic_yang_models-1.0-py3-none-any.whl`
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x ] 202012

### CODE REVIEW

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/feature.json [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                        "has_global_scope": true,
+                        "has_asic_scope": false
+                    },
+                    {
+                        "name": "lldp",
+                        "state": "disabled",
+                        "auto_restart": "disabled",
+                        "has_timer": false,
+                        "has_global_scope": false,
+                        "has_asic_scope": true
```
Comment (arlakshm) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/feature.json:44]: check the `status` field in minigraph.xml ?

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/feature.json [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                    }
+                ]
+            }
+        }
+    },
+    "FEATURE_WITH_INVALID_STATE": {
```
Comment (arlakshm) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/feature.json:50]: check with tests with hostcfgd as well

File: src/sonic-yang-models/tests/files/sample_config_db.json [hunk 3]
```diff
@@ -907,6 +907,56 @@
...
+                "high_mem_alert": "disabled",
+                "state": "enabled"
+            },
+            "syncd": {
+                "auto_restart": "enabled",
+                "has_global_scope": "False",
+                "has_per_asic_scope": "True",
+                "has_timer": "False",
+                "high_mem_alert": "disabled",
+                "state": "enabled"
```
Comment (arlakshm) [inline src/sonic-yang-models/tests/files/sample_config_db.json:950]: check for telemetry https://github.com/Azure/sonic-buildimage/blob/9d18a35e35d3f57c00ba14cae87568a7a203d5d6/src/sonic-config-engine/minigraph.py#L1549

**Comment by zhangyanzhao** [comment]
not ready for review, need move it back to "To be Reviewed column"

File: src/sonic-yang-models/yang-models/sonic-feature.yang [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                }
+
+                leaf auto_restart {
+                    type stypes:feature_state;
+
+                }
+
+                leaf has_timer {
+                    description "This configuration identicates if there is
+                                timer associated to this feature";
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-feature.yang:43]: what's the purpose of this timer? time to start the feature?

File: src/sonic-yang-models/yang-models/sonic-feature.yang [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                    type boolean;
+                    default false;
+                }
+
+                leaf high_mem_alert {
+                    description "This configuration controls the trigger to generate
+                                alert on high memory utilization";
+                    type stypes:feature_state;
+                    default "disabled";
+                }
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-feature.yang:67]: no field for high_cpu_alert?

Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-feature.yang:67]: I remember, there was a file for CPU & memory allocation per feature/docker, is that not exposed to the user yet?

File: src/sonic-yang-models/yang-models/sonic-feature.yang [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            description "feature table in config_db.json";
+
+            list FEATURE_LIST {
+
+                key "name";
+
+                leaf name {
+                    description "feature name in Feature table";
+                    type string {
+                        length 1..32;
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-feature.yang:27]: Any reason for restricting the len to 32?

File: src/sonic-yang-models/yang-models/sonic-feature.yang [hunk 7]
```diff
@@ -0,0 +1,71 @@
+module sonic-feature{
+
+    yang-version 1.1;
+
+    namespace "http://github.com/Azure/sonic-feature";
+    prefix port;
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-feature.yang:6]: Please fix the prefix.

File: src/sonic-yang-models/yang-models/sonic-feature.yang [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+                leaf name {
+                    description "feature name in Feature table";
+                    type string {
+                        length 1..32;
+                    }
+                }
+
+                leaf state {
+                    description "state of the feature";
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-feature.yang:32]: Dont we have any default value? or is this not a mandatory field?

**Comment by zhangyanzhao** [comment]
Arvind will handle the comments from Venkat.

**Comment by venkatmahalingam** [comment]
@arlakshm In the mgmt VRF YANG, boolean value is represented as "true" in the below file, please try this for your case as well.
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/tests/files/sample_config_db.json#L281

**Comment by zhangyanzhao** [comment]
@arlakshm  please address the comments from Venkat. Thanks.

File: src/sonic-yang-models/yang-models/sonic-feature.yang (hunk 4 above)
Comment (arlakshm) [reply src/sonic-yang-models/yang-models/sonic-feature.yang:43]: Some the features have timer associated. `snmp` is an example. This attributed to indicate that field

File: src/sonic-yang-models/yang-models/sonic-feature.yang (hunk 6 above)
Comment (arlakshm) [reply src/sonic-yang-models/yang-models/sonic-feature.yang:27]: I wanted to put a max length for the feature name. Thirty-two would be enough to cover all existing features. If we need a bigger string, we can modify in the future.

File: src/sonic-yang-models/yang-models/sonic-feature.yang (hunk 5 above)
Comment (arlakshm) [reply src/sonic-yang-models/yang-models/sonic-feature.yang:67]: currently there is no support for `high_cpu_alert `now. We update once the feature is supported

File: src/sonic-yang-models/yang-models/sonic-feature.yang (hunk 7 above)
Comment (arlakshm) [reply src/sonic-yang-models/yang-models/sonic-feature.yang]: Fixed in the latest commit

File: src/sonic-yang-models/yang-models/sonic-feature.yang (hunk 8 above)
Comment (arlakshm) [reply src/sonic-yang-models/yang-models/sonic-feature.yang:32]: Add a default value in latest commit

**Comment by arlakshm** [comment]
> @arlakshm In the mgmt VRF YANG, boolean value is represented as "true" in the below file, please try this for your case as well.
> https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/tests/files/sample_config_db.json#L281

The backend python code expects "True" or "False"  I think change to "true" may not work.

**Comment by venkatmahalingam** [comment]
> > @arlakshm In the mgmt VRF YANG, boolean value is represented as "true" in the below file, please try this for your case as well.
> > https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/tests/files/sample_config_db.json#L281
> 
> The backend python code expects "True" or "False" I think change to "true" may not work.

boolean type is represented as "true/"false" in the DB for tables (MGMT VRF, LLDP..etc), can we change the backend to accept "true"/"false" to be consistent.

**Review by venkatmahalingam — APPROVED** [review]

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/feature.json (hunk 1 above)
Comment (arlakshm) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/feature.json]: The Telemetry feature also has the same state as other feature
https://github.com/Azure/sonic-buildimage/blob/b3b6938fda9244607fb00bfd36a74bccab0c38a9/files/build_templates/init_cfg.json.j2#L39

File: src/sonic-yang-models/tests/files/sample_config_db.json (hunk 3 above)
Comment (arlakshm) [reply src/sonic-yang-models/tests/files/sample_config_db.json:1111]: The Telemetry feature also has the same state as other feature
https://github.com/Azure/sonic-buildimage/blob/b3b6938fda9244607fb00bfd36a74bccab0c38a9/files/build_templates/init_cfg.json.j2#L39

File: src/sonic-yang-models/yang-models/sonic-feature.yang [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+                    description "This configuration identicates there will only one service
+                                spawned per asic";
+                    type boolean;
+                    default false;
+                }
+
+                leaf high_mem_alert {
+                    description "This configuration controls the trigger to generate
+                                alert on high memory utilization";
+                    type stypes:feature_state;
```
Comment (abdosi) [inline src/sonic-yang-models/yang-models/sonic-feature.yang:66]: @arlakshm does always_enabled applies for high_mem_alert ?

Comment (yozhao101) [reply src/sonic-yang-models/yang-models/sonic-feature.yang:66]: > @arlakshm does always_enabled applies for high_mem_alert ?

I think it is OK to let `high_mem_alert` support the option `always_enabled`.

**Review by abdosi — APPROVED** [review]

**Pull request merged**